File: vhdl_toolkit/vhdlConvertor/hdlObjects/Context.java
package hdlObjects;

import java.util.List;
import java.util.Vector;

import org.json.JSONException;
import org.json.JSONObject;

public class Context extends Jsonable {
	public List<Expr> imports;
	public List<Entity> entities;
	public List<Arch> architectures;
	public List<Package> packages;
	public List<PackageHeader> packageHeaders;

	public Context() {
		imports = new Vector<Expr>();
		entities = new Vector<Entity>();
		architectures = new Vector<Arch>();
		packages = new Vector<Package>();
		packageHeaders = new Vector<PackageHeader>();
	}

	public JSONObject toJson() throws JSONException {
		JSONObject c = new JSONObject();
		addJsonArr(c, "imports", imports);
		addJsonObj(c, "entities", entities, e -> e.name);
		addJsonArr(c, "architectures", architectures);
		addJsonObj(c, "packages", packages, p -> p.name);
		addJsonObj(c, "packageHeaders", packageHeaders, ph -> ph.name);
		return c;
	}

}

File: hdl_toolkit/parser/hdlConvertor/src/hdlObjects/context.cpp
#include "context.h"

PyObject * Context::toJson() const {
	PyObject * c = PyDict_New();
	addJsonArrP(c, "imports", imports);
	addJsonArrP(c, "entities", entities);
	addJsonArrP(c, "architectures", architectures);
	addJsonArrP(c, "packages", packages);
	addJsonArrP(c, "packageHeaders", packageHeaders);
	Py_IncRef(c);
	return c;
}

void Context::dump(int indent) const {
	mkIndent(indent) << "{\n";
	indent += INDENT_INCR;
	dumpArrP("imports", indent, imports) << ",\n";
	dumpArrP("entities", indent, entities) << ",\n";
	dumpArrP("architectures", indent, architectures) << ",\n";
	dumpArrP("packages", indent, packages) << ",\n";
	dumpArrP("packageHeaders", indent, packageHeaders) << ",\n";
	indent -= INDENT_INCR;
	mkIndent(indent) << "}";
}

Context::~Context() {
	for (auto i : imports) {
		delete i;
	}
	for (auto i : entities) {
		delete i;
	}
	for (auto i : architectures) {
		delete i;
	}
	for (auto i : packages) {
		delete i;
	}
	for (auto i : packageHeaders) {
		delete i;
	}
}
